### PR TITLE
Qemu helpers & hooks refactoring

### PR DESCRIPTION
**Pull request by rmalmain**
It's a first draft addressing issue #1985 and PR #1784.

Main changes of this PR: 
- Rename **Helper** into **Tool**, so `QemuHelper` becomes `EmulatorTool`, `QemuHelperTuple` becomes `EmulatorToolTuple`, etc...
- Hooks have been separated into a low-level and high-level part, like for QEMU: `qemu/hooks.rs` contains the low-level (stateless in the rust part) struct `QemuHooks` and `emu/hooks.rs` the low-level (with a rust state) counterpart, `EmulatorHooks` and `EmulatorTools`.
- The high-level static hook variables have been centralized in the `EmulatorHooks` struct.
- A lot of hook-related types now have dedicated names for readability (check the `create_hook_types` macro)
- `EmulatorTools` owns `EmulatorToolTuple`, and `Emulator` owns `EmulatorTools`.
- The Qemu Executor now contains a ref to `Emulator`, not `QemuHooks`. The Stateful Executor now gives access to `Emulator` as the second argument of the harness (still through the state, I need to change this) (check the `qemu_systemmode` classic example to see how it changes)
- And a lot of smaller changes I probably forgot.

What remains to do:
- [ ] We still need to decide on a fitting name for `Emulator` that represents the fact it's a high level abstraction of `Qemu`.
- [ ] Not sure if we should call the `EmulatorToolTuple` callbacks in the executor or in the emulator run function. For now I kept it in Executor as it was before, we can easily move this now since `Emulator` owns the tuple.
- [x] Change the `E` bound to `EH` bound for exit handlers to avoid the collision with the Executor generic.
- [x] Change the `QT` bound to `ET` (or `ETT`?) to keep consistent with `EmulatorToolTuple`. Didn't do it yet to confirm `EmulatorTool` is a correct name.
- [ ] Should we make `StatefulQemuExecutor` the new default `QemuExecutor` with this new organization? Since it gives access to `Emulator`, it sounds like something we always want, even if `Emulator` is an empty shell only instantiated to use QEMU low-level API.
- [ ] Revamp of the `Emulator` builder. `TypedBuilder` has some flaws that make it unusable with generic members. We now need it to work correctly since `Emulator` will most likely be used everywhere.

### CODE REVIEW

**Comment by andreafioraldi** [comment]
i guess https://github.com/AFLplusplus/LibAFL/pull/1784 is superseded by this, closing

File: fuzzers/qemu_systemmode/src/fuzzer_breakpoint.rs [hunk 1]
```diff
@@ -30,12 +30,11 @@ use libafl_bolts::{
 use libafl_qemu::{
     breakpoint::Breakpoint,
     command::{EndCommand, StartCommand, StdCommandManager},
-    edges::{edges_map_mut_ptr, QemuEdgeCoverageHelper, EDGES_MAP_SIZE_IN_USE, MAX_EDGES_FOUND},
+    edges::{edges_map_mut_ptr, QemuEdgeCoverageTool, EDGES_MAP_SIZE_IN_USE, MAX_EDGES_FOUND},
```
Comment (domenukk) [inline fuzzers/qemu_systemmode/src/fuzzer_breakpoint.rs:33]: Frida uses `CoverageRuntime`

Comment (domenukk) [reply fuzzers/qemu_systemmode/src/fuzzer_breakpoint.rs:33]: (and it's not really a tool, is it?)

Comment (domenukk) [reply fuzzers/qemu_systemmode/src/fuzzer_breakpoint.rs:33]: Maybe `Plugin`/`Extension`?

Comment (rmalmain) [reply fuzzers/qemu_systemmode/src/fuzzer_breakpoint.rs:33]: `Plugin` collides with qemu plugins, it could be misunderstood no?

Comment (domenukk) [reply fuzzers/qemu_systemmode/src/fuzzer_breakpoint.rs:33]: We should keep the names for the same usecase the same between Frida and QEMU though, I think

Comment (domenukk) [reply fuzzers/qemu_systemmode/src/fuzzer_breakpoint.rs:33]: We can also change the frida name I guess

Comment (domenukk) [reply fuzzers/qemu_systemmode/src/fuzzer_breakpoint.rs:33]: Maybe ask ChatGPT :D

File: fuzzers/qemu_coverage/src/fuzzer.rs [hunk 2]
```diff
@@ -121,28 +119,11 @@ pub fn fuzz() {
...
-        .fold(
-            RangeMap::<usize, (u16, String)>::new(),
-            |mut rm, (i, (r, p))| {
-                rm.insert(r, (i as u16, p));
-                rm
-            },
-        );
-
-    let mut coverage = PathBuf::from(&options.coverage);
+    let mut coverage_path = PathBuf::from(&options.coverage);
```
Comment (domenukk) [inline fuzzers/qemu_coverage/src/fuzzer.rs:122]: The option should directly be a `PathBuf` and `into`

File: libafl_sugar/src/qemu.rs [hunk 3]
```diff
@@ -222,19 +226,26 @@ where
             };
 
             if self.use_cmplog.unwrap_or(false) {
-                let mut hooks = QemuHooks::new(
-                    *qemu,
+                let tools = {
```
Comment (domenukk) [inline libafl_sugar/src/qemu.rs:229]: `let mods`

File: libafl_qemu/src/modules/usermode/snapshot.rs [hunk 4]
```diff
@@ -796,27 +810,37 @@ where
         )))]
         SYS_newfstatat => {
             if a2 != 0 {
-                let h = hooks.match_helper_mut::<QemuSnapshotHelper>().unwrap();
+                let h = emulator_modules
+                    .match_module_mut::<SnapshotModule>()
```
Comment (domenukk) [inline libafl_qemu/src/modules/usermode/snapshot.rs:814]: `get_mut`? It's always going to be a module, right?

Comment (rmalmain) [reply libafl_qemu/src/modules/usermode/snapshot.rs:814]: i'd still keep the match part since it will return the first match of the module tuple

Comment (rmalmain) [reply libafl_qemu/src/modules/usermode/snapshot.rs:814]: nevermind, we would have to use `match()`, which is a reserved keyword

**Review by domenukk — APPROVED** [review]

**Comment by rmalmain** [comment]
i'll merge the PR in its current state. There could be other things to add, but i believe it's already big enough like this.
let's address the other points in separate PRs.

**Pull request merged**